lib/charms/mysql/v0/mysql.py

```diff
@@ -535,36 +540,43 @@ def has_cos_relation(self) -> bool:
 
         return len(active_cos_relations) > 0
 
-    def _get_secret_from_juju(self, scope: str, key: str) -> Optional[str]:
-        """Retrieve and return the secret from the juju secret storage."""
-        if scope == "unit":
-            secret_id = self.unit_peer_data.get(SECRET_ID_KEY)
+    def _scope_obj(self, scope: Scopes):
+        if scope == APP_SCOPE:
+            return self.app
+        if scope == UNIT_SCOPE:
+            return self.unit
 
-            if not self.unit_secrets and not secret_id:
-                logger.debug("Getting a secret when no secrets added in juju")
-                return None
-
-            if not self.unit_secrets:
-                secret = self.model.get_secret(id=secret_id)
-                content = secret.get_content()
-                self.unit_secrets = content
-                logger.debug(f"Retrieved secret {key} for unit from juju")
+    def _peer_data(self, scope: Scopes) -> Dict:
+        """Return corresponding databag for app/unit."""
+        if self.peers is None:
+            return {}
+        return self.peers.data[self._scope_obj(scope)]
 
-            return self.unit_secrets.get(key)
+    def _safe_get_secret(self, scope: Scopes, label: str) -> SecretCache:
+        """Safety measure, for upgrades between versions based on secret URI usage to others with labels usage.
+        If the secret can't be retrieved by label, we search for the uri -- and if found, we "stick" the
+        label on the secret for further usage.
+        """
+        secret_uri = self._peer_data(scope).get(SECRET_ID_KEY, None)
+        secret = self.secrets.get(label, secret_uri)
 
-        secret_id = self.app_peer_data.get(SECRET_ID_KEY)
+        # Since now we switched to labels, the databag reference can be removed
+        if secret_uri and secret and scope == APP_SCOPE and self.unit.is_leader():
+            self._peer_data(scope).pop(SECRET_ID_KEY, None)
+        return secret
 
-        if not self.app_secrets and not secret_id:
-            logger.debug("Getting a secret when no secrets added in juju")
-            return None
+    def _get_secret_from_juju(self, scope: Scopes, key: str) -> Optional[str]:
+        """Retrieve and return the secret from the juju secret storage."""
+        label = generate_secret_label(self, scope)
+        secret = self._safe_get_secret(scope, label)
 
-        if not self.app_secrets:
-            secret = self.model.get_secret(id=secret_id)
-            content = secret.get_content()
-            self.app_secrets = content
-            logger.debug(f"Retrieved secret {key} for app from juju")
+        if not secret:
+            logger.debug("Getting a secret when secret is not added in juju")
+            return
 
-        return self.app_secrets.get(key)
+        value = secret.get_content().get(key)
+        logger.debug(f"Retrieved secret {key} for unit from juju")
```

Comment:
can we remove this debug statement? as i understand it, it will log every time `_get_secret_from_juju` is called, which is going to be a lot